### PR TITLE
feat(program): add replaceSource method

### PR DESCRIPTION
**Pull request by yjhmelody**
It's a convinient method for user to update a Source.

### CODE REVIEW

**Comment by willemneal** [comment]
This is failing because it uses closures and thus can't be bootstrapped.

```ts
ERROR AS100: Not implemented: Closures

       (source) => source.internalPath !== newSource.internalPath
```

**Comment by yjhmelody** [comment]
> This is failing because it uses closures and thus can't be bootstrapped.
> 
> ```ts
> ERROR AS100: Not implemented: Closures
> 
>        (source) => source.internalPath !== newSource.internalPath
> ```

Well, I will fix it. BTW, how to add a unit test to test this API.

File: src/program.ts [hunk 1]
```diff
@@ -944,6 +944,34 @@ export class Program extends DiagnosticEmitter {
...
+    var parser = this.parser;
+    parser.donelog.delete(newSource.internalPath);
+    parser.seenlog.delete(newSource.internalPath);
+    var sources = parser.sources;
+    var newSources: Source[] = [];
+    for(let i = 0; i < sources.length; i++) {
+      if(sources[i].internalPath != newSource.internalPath) {
+        newSources.push(sources[i]);
+      }
+    }
```
Comment (MaxGraey) [inline src/program.ts:958]:
```suggestion
    var newSources: Source[] = [];
    for (let i = 0, len = sources.length; i < len; i++) {
      if (sources[i].internalPath != newSource.internalPath) {
        newSources.push(sources[i]);
      }
    }
```

File: src/program.ts [hunk 2]
```diff
@@ -944,6 +944,34 @@ export class Program extends DiagnosticEmitter {
...
+    parser.sources = newSources;
+    sources = this.sources;
+    newSources = [];
+    for(let i = 0; i < sources.length; i++) {
+      if(sources[i].internalPath != newSource.internalPath) {
+        newSources.push(sources[i]);
+      }
+    }
```
Comment (MaxGraey) [inline src/program.ts:966]:
```suggestion
    for (let i = 0, len = sources.length; i < len; i++) {
      if (sources[i].internalPath != newSource.internalPath) {
        newSources.push(sources[i]);
      }
    }
```

File: src/program.ts [hunk 3]
```diff
@@ -944,6 +944,34 @@ export class Program extends DiagnosticEmitter {
       : null;
   }
 
+  /** Replace a old source with a new source according to internalPath */
+  replaceSource(newSource: Source): void {
+    var parser = this.parser;
+    parser.donelog.delete(newSource.internalPath);
+    parser.seenlog.delete(newSource.internalPath);
```
Comment (yjhmelody) [inline src/program.ts:951]: Is the `delete` meaningful ? I found that near-sdk will do such an operation, but I'm not sure it's neccessary.

Comment (MaxGraey) [reply src/program.ts:951]: cc @willemneal

Comment (MaxGraey) [reply src/program.ts:951]: At first glance such delete and add same paths looks unnecessary

**Comment by MaxGraey** [comment]
Also it looks like `this.sources` and `this.parser.sources` should be the same so update sources twice for both arrays looks like unnecessary. But I may miss something

**Comment by MaxGraey** [comment]
Basically I guess code may be look very simple:
```ts
replaceSource(newSource: Source): void {
  var parser = this.parser;
  var sources = parser.sources;
  for (let i = 0, len = sources.length; i < len; i++) {
    if (sources[i].internalPath == newSource.internalPath) {
      sources[i] = newSource;
      break;
    }
  }
}
```
that's it

**Comment by yjhmelody** [comment]
> Also it looks like `this.sources` and `this.parser.sources` should be the same reference so update sources twice for this both arrays looks like unnecessary. But I may miss something but on rest of code only `this.sources` used so I suggest update `this.parser.sources` only and reassign `this.sources = this.parser.sources` afterwards in needed

parser.sources is not program.sources.

**Comment by yjhmelody** [comment]
> Basically I guess code may be look very simple:
> 
> ```ts
> replaceSource(newSource: Source): void {
>   var parser = this.parser;
>   var sources = parser.sources;
>   for (let i = 0, len = sources.length; i < len; i++) {
>     if (sources[i].internalPath == newSource.internalPath) {
>       sources[i] = newSource;
>       break;
>     }
>   }
>   this.sources = sources;
> }
> ```
> 
> that's it

You are right. I made it complicated.

**Comment by MaxGraey** [comment]
Also plz add test for `replaceSource` here: https://github.com/AssemblyScript/assemblyscript/blob/master/tests/browser-asc.js

**Comment by yjhmelody** [comment]
> Also plz add test for `replaceSource` here: https://github.com/AssemblyScript/assemblyscript/blob/master/tests/browser-asc.js

Sorry, I don’t quite understand what you said, it seems that this file cannot test program-related APIs

File: tests/browser-asc.js [hunk 4]
```diff
@@ -99,6 +100,8 @@ newParser.parseFile(newCode, "index2.ts", true);
 const internalPath = parser.sources[0].internalPath;
 const newCodeSource = newParser.sources[0];
 newCodeSource.internalPath = internalPath;
-
 program.replaceSource(newCodeSource);
-console.log(`After replaceSource: ${parser.sources[0].statements[0]}`);
\ No newline at end of file
+console.log(`After replaceSource: ${parser.sources[0].statements[0]}`);
+if (parser.sources[0] != newCodeSource) {
+  throw new
```
Comment (MaxGraey) [inline tests/browser-asc.js]:
```suggestion
  throw new Error("source code have not been updated after calling replaceSource!");
```

File: src/program.ts [hunk 5]
```diff
@@ -944,6 +944,18 @@ export class Program extends DiagnosticEmitter {
       : null;
   }
 
+  /** Replace a old source with a new source according to internalPath */
```
Comment (MaxGraey) [inline src/program.ts:947]:
```suggestion
  /** Replace an old source with a new source according to internalPath */
```

File: tests/browser-asc.js [hunk 6]
```diff
@@ -79,5 +85,23 @@ process.stdout.write(output.stdout.toString());
 console.log(">>> .stderr >>>");
 process.stdout.write(output.stderr.toString());
 console.log(">>> .text >>>");
-process.stdout.write(output.text);
-console.log(">>> .binary >>> " + output.binary.length + " bytes");
+// process.stdout.write(output.text);
+// console.log(">>> .binary >>> " + output.binary.length + " bytes");
```
Comment (MaxGraey) [inline tests/browser-asc.js:89]: why this commented?

Comment (yjhmelody) [reply tests/browser-asc.js:89]: I accidentally commented, because it can't run locally

**Comment by yjhmelody** [comment]
@MaxGraey  Anything need to be changed?

**Pull request closed without merging**